File: leetcode-solutions/701. Insert into a Binary Search Tree.rb
# https://leetcode.com/problems/insert-into-a-binary-search-tree/

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} val
# @return {TreeNode}
def insert_into_bst(root, val)
  node = root
  while node 
      if node.val > val
          unless node.left
              node.left = TreeNode.new(val)
              break 
          end
          node = node.left
      else
          unless node.right
              node.right = TreeNode.new(val)
              break 
          end
          node = node.right
      end
  end
  
  root
end

File: leetcode-solutions/python/701. Insert into a Binary Search Tree.py
# https://leetcode.com/problems/insert-into-a-binary-search-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if root is None: return TreeNode(val)
        current = root
        while current:
            if current.val > val:
                if current.left is None:
                    current.left = TreeNode(val)
                    return root
                current = current.left
            else:
                if current.right is None:
                    current.right = TreeNode(val)
                    return root
                current = current.right
        return root
